### ISSUE TITLE
more detail about the function _l2_loss() and no second-order approximation?

### ISSUE THREAD

**Issue by ee-hash** [xujinfan/NASP-codes#8]
Hi!
I 'm confused about the way to achieve no second-order approximation?  
In the model_search.py and architect.py, like this. It has update the parameter A. 
 `
  >self.model.binarization()
  >loss = self.model._loss(input_valid, target_valid, updateType)
  >loss.backward()
  >self.model.restore()
  >self.optimizer.step()`

I have two confusion:
1. when updating A, just convert all weights of operation to one-hot vector, the weight of selected operation will be 1, and other be 0. Then, update A without second-order approximation. Is it optimize right? 
2. Could you introduce the L2_loss? I I don't quite understand these  mathematical formula in section A.1.

Thanks! Hope your reply!

**Comment by xujinfan** [comment]
Hello，you can read the paper “binary connect” to understand it.

**Comment by hwx10** [comment]
> Hello，you can read the paper “binary connect” to understand it.
> […](#)
> 在 2021年4月28日，20:13，hwx ***@***.***> 写道： ﻿ Hi! I 'm confused about the way to achieve no second-order approximation? In the model_search.py and architect.py, like this. It has update the parameter a. `>self.model.binarization() loss = self.model._loss(input_valid, target_valid, updateType) loss.backward() self.model.restore() self.optimizer.step()` I have two confusion: when updating A, just convert all weights of operation to one-hot vector, the weight of selected operation will be 1, and other be 0. Then, update A without second-order approximation. Is it optimize right? Could you introduce the L2_loss? I I don't quite understand these mathematical formula in section A.1. Thanks! — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or unsubscribe.

Thanks for you reply.
I have read the paper--binary connect. You mean that binarization could substitute  second-order approximation?
Then, could you explain more detail of the L2 loss code?
`
>  def _l2_loss(self):
>    normal_burden = []
>    params = 0
>    for key in PRIMITIVES_NORMAL:
>      params += PARAMS[key]
>    for key in PRIMITIVES_NORMAL:
>      normal_burden.append(PARAMS[key]/params)
>    normal_burden = torch.autograd.Variable(torch.Tensor(normal_burden).cuda(), requires_grad=False)
>    return (self.alphas_normal*self.alphas_normal*normal_burden).sum()*self._l2
` 
Thanks!
